kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -112,91 +126,94 @@ private static String generateAsyncThreadName(
    */
   public <KOut, VOut> void scheduleForProcessing(
       final String processorName,
-      final int taskId,
+      final int partition,
       final List<AsyncEvent> events,
       final FinalizingQueue finalizingQueue,
       final ProcessingContext taskContext,
       final AsyncUserProcessorContext<KOut, VOut> asyncProcessorContext
   ) {
-    // todo: can also check in-flight for failed tasks
+    final var inFlightKey = InFlightWorkKey.of(processorName, partition);
+    final var inFlightForTask
+        = inFlight.computeIfAbsent(inFlightKey, k -> new ConcurrentHashMap<>());
     for (final var e : events) {
-      final var inFlightKey = InFlightWorkKey.of(processorName, taskId);
-      inFlight.computeIfAbsent(inFlightKey, k -> new ConcurrentHashMap<>());
-      final var inFlightForTask = inFlight.get(inFlightKey);
-      if (inFlightForTask.isEmpty()) {
-        log.debug("added in-flight map for task {}", taskId);
-      }
-      // hold the lock for the in-flight records so that we ensure that the future is inserted
-      // before it is removed by the task
-      synchronized (inFlightForTask) {
-        final Future<?> future = executorService.submit(new AsyncEventTask<>(
-            e,
-            taskContext,
-            asyncProcessorContext,
-            finalizingQueue,
-            inFlightForTask,
-            log));
-        inFlightForTask.put(e, future);
-      }
+      final var future = CompletableFuture.runAsync(
+          new AsyncEventTask<>(
+              e, taskContext, asyncProcessorContext, finalizingQueue, inFlightForTask),
+          executor
+      );
+      final var inFlightEvent = new InFlightEvent(future);
+      inFlightForTask.put(e, inFlightEvent);
+      future.handle((r, t) -> {
+        if (t != null) {
+          inFlightEvent.setError(t);
+        } else {
+          inFlightForTask.remove(e);
```

Comment:
I guess it doesn't really matter right now, but looking forward to when we might want to allow actually handling processing exceptions and need to keep the internal state consistent even after a failure...we should probably remove it from the inFlight map regardless of whether it failed?

Reply:
Not in this case. This is really just a failsafe. Any exception that we want to handle gracefully should go onto the finalizing queue. The check here is just to catch anything that slips through, and if that happens its a bug and we should exit the main thread (when it calls `checkUncaughtExceptions`)